### PR TITLE
Merge in hash and parsing improvements

### PR DESCRIPTION
**Pull request by Ericson2314**
This will help with the new content address parsers we need to write, and while these changes are not yet merged to master I think they might just be soon.

### CODE REVIEW

File: src/libfetchers/fetchers.cc [hunk 1]
```diff
@@ -36,8 +36,7 @@ std::unique_ptr<Input> inputFromAttrs(const Attrs & attrs)
         auto res = inputScheme->inputFromAttrs(attrs2);
         if (res) {
             if (auto narHash = maybeGetStrAttr(attrs, "narHash"))
-                // FIXME: require SRI hash.
-                res->narHash = newHashAllowEmpty(*narHash, {});
+                res->narHash = Hash::parseSRI(*narHash);
```
Comment (matthewbauer) [inline src/libfetchers/fetchers.cc:39]: @Ericson2314 does parseSRI handle the empty case?

Comment (Ericson2314) [reply src/libfetchers/fetchers.cc:39]: This never in fact worked in the empty case, because there is no hash type.

Comment (matthewbauer) [reply src/libfetchers/fetchers.cc:39]: ok

Comment (matthewbauer) [reply src/libfetchers/fetchers.cc:39]: we should have a way to do this. maybe through `narHash = "sha256-"`

Comment (Ericson2314) [reply src/libfetchers/fetchers.cc:39]: Yeah that makes sense to me.

**Pull request merged**